module/zfs/dmu_send.c

```diff
@@ -507,11 +543,27 @@ dump_dnode(dmu_sendarg_t *dsp, uint64_t object, dnode_phys_t *dnp)
 	    drro->drr_blksz > SPA_OLD_MAXBLOCKSIZE)
 		drro->drr_blksz = SPA_OLD_MAXBLOCKSIZE;
 
-	if (dump_record(dsp, DN_BONUS(dnp),
-	    P2ROUNDUP(dnp->dn_bonuslen, 8)) != 0) {
-		return (SET_ERROR(EINTR));
+	if ((dsp->dsa_featureflags & DMU_BACKUP_FEATURE_RAW) &&
+	    BP_IS_PROTECTED(bp)) {
+		drro->drr_flags |= DRR_RAW_ENCRYPTED;
+		if (BP_SHOULD_BYTESWAP(bp))
+			drro->drr_flags |= DRR_RAW_BYTESWAP;
+
+		/* needed for reconstructing dnp on recv side */
+		drro->drr_indblkshift = dnp->dn_indblkshift;
+		drro->drr_nlevels = dnp->dn_nlevels;
+		drro->drr_nblkptr = dnp->dn_nblkptr;
+
+		/* raw bonus buffers extend to the end of the dnp */
+		if (bonuslen != 0) {
+			drro->drr_raw_bonuslen = DN_MAX_BONUS_LEN(dnp);
```

Comment:
Can you explain *why* for raw sends, the entire bonus area needs to be sent, rather than just the part that's filled in (dn_bonuslen)?  The difference should be all zeros, so there shouldn't be any point to sending it (but maybe it isn't all zeros because....?)